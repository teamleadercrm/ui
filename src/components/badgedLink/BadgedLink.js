import uiTypography from '@teamleader/ui-typography';
import uiUtilities from '@teamleader/ui-utilities';
import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { forwardRef, useImperativeHandle, useRef } from 'react';
import Box from '../box';
import theme from './theme.css';

const BadgedLink = forwardRef(
  (
    {
      children,
      className,
      disabled,
      icon,
      iconPlacement,
      element,
      inherit,
      iconAligment,
      inverse,
      selected,
      onMouseUp,
      onMouseLeave,
      ...others
    },
    ref,
  ) => {
    const badgedLinkRef = useRef();
    useImperativeHandle(ref, () => badgedLinkRef.current);

    const blur = () => {
      const currentBadgedLinkRef = badgedLinkRef.current;
      if (currentBadgedLinkRef.blur) {
        currentBadgedLinkRef.blur();
      }
    };

    const handleMouseUp = (event) => {
      blur();
      onMouseUp && onMouseUp(event);
    };

    const handleMouseLeave = (event) => {
      blur();
      onMouseLeave && onMouseLeave(event);
    };

    const classNames = cx(
      uiUtilities['reset-font-smoothing'],
      uiTypography['ui-text-body'],
      theme['badged-link'],
      {
        [theme['is-inherit']]: inherit,
        [theme['is-bottom-aligned']]: iconAligment === 'bottom',
        // Since children can be text node, we have to do spacing according to icon element
        [theme['is-left-icon-placed']]: iconPlacement === 'left',
      },
      className,
    );

    return (
      <Box
        element={element}
        ref={badgedLinkRef}
        className={classNames}
        data-teamleader-ui="badgedLink"
        onMouseUp={handleMouseUp}
        onMouseLeave={handleMouseLeave}
        {...others}
      >
        {icon && iconPlacement === 'left' && icon}
        {children}
        {icon && iconPlacement === 'right' && icon}
      </Box>
    );
  },
);

BadgedLink.propTypes = {
  /** The content to display inside the badged link. */
  children: PropTypes.any.isRequired,
  /** A class name for the badged link to give custom styles. */
  className: PropTypes.string,
  /** The icon displayed inside the button. */
  icon: PropTypes.element.isRequired,
  /** The aligment of the icon inside the button. */
  iconAligment: PropTypes.oneOf(['center', 'bottom']),
  /** The position of the icon inside the button. */
  iconPlacement: PropTypes.oneOf(['left', 'right']),
  /** If true, the badged link style inherits the parent element style. */
  inherit: PropTypes.bool,
  /** A custom element to be rendered */
  element: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  /** Callback function that is fired when mouse leaves the component. */
  onMouseLeave: PropTypes.func,
  /** Callback function that is fired when the mouse button is released. */
  onMouseUp: PropTypes.func,
};

BadgedLink.defaultProps = {
  className: '',
  element: 'a',
  iconAligment: 'center',
  iconPlacement: 'left',
  inherit: true,
};

BadgedLink.displayName = 'BadgedLink';

export default BadgedLink;
