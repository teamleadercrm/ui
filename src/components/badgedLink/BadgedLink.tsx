import uiTypography from '@teamleader/ui-typography';
import uiUtilities from '@teamleader/ui-utilities';
import cx from 'classnames';
import React, { ElementType, forwardRef, ReactNode } from 'react';
import { GenericComponent } from '../../@types/types';
import Box from '../box';
import { BoxProps } from '../box/Box';
import theme from './theme.css';

export interface BadgedLinkProps extends Omit<BoxProps, 'ref'> {
  /** The content to display inside the badged link. */
  children: ReactNode;
  /** A class name for the badged link to give custom styles. */
  className?: string;
  /** The icon displayed inside the button. */
  icon?: ReactNode;
  /** The position of the icon inside the button. */
  iconPlacement?: 'left' | 'right';
  /** If true, the badged link style inherits the parent element style. */
  inherit?: boolean;
  /** A custom element to be rendered */
  element?: ElementType | undefined;
}

const BadgedLink: GenericComponent<BadgedLinkProps> = forwardRef<HTMLElement, BadgedLinkProps>(
  ({ children, className = '', icon, iconPlacement = 'left', element = 'a', inherit = true, ...others }, ref) => {
    const classNames = cx(
      uiUtilities['reset-font-smoothing'],
      uiTypography['ui-text-body'],
      theme['badged-link'],
      {
        [theme['is-inherit']]: inherit,
        // Since children can be text node, we have to do spacing according to icon element
        [theme['is-left-icon-placed']]: iconPlacement === 'left',
      },
      className,
    );

    return (
      <Box element={element} ref={ref} className={classNames} data-teamleader-ui="badgedLink" {...others}>
        {iconPlacement === 'left' && icon}
        {children}
        {iconPlacement === 'right' && icon}
      </Box>
    );
  },
);

BadgedLink.displayName = 'BadgedLink';

export default BadgedLink;
