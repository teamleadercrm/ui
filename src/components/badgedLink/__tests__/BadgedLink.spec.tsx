import React from 'react';
import { render } from '@testing-library/react';
import BadgedLink from '../BadgedLink';

describe('Component - BadgedLink', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <BadgedLink href="https://foobar.com" target="_blank">
        Foobar
      </BadgedLink>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
