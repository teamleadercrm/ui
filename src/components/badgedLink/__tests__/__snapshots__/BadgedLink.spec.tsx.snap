// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - BadgedLink renders 1`] = `
<DocumentFragment>
  <a
    class="box reset-font-smoothing ui-text-body badged-link is-inherit is-left-icon-placed"
    data-teamleader-ui="badgedLink"
    href="https://foobar.com"
    target="_blank"
  >
    Foobar
  </a>
</DocumentFragment>
`;
